src/Dijkstra/AST/Maybe.agda

```diff
@@ -169,6 +176,21 @@ predTrans-is-weakest (ASTbind m f) _ Pr
 ... | just x  = rec _ λ where r refl → predTrans-is-weakest (f x) _ Pr
 predTrans-is-weakest (ASTop Maybe-bail f) P = id
 
+maybePTBindLemma : ∀ {A B : Set} {m : MaybeD A} {f : A → MaybeD B} {P : Post B}{i : Input}
+                   → (prog : MaybeD B)
+                   → prog ≡ ASTbind m f
+                   → (      runMaybe m i ≡ nothing → P nothing)
+                   → (∀ x → runMaybe m i ≡ just x  → P (runMaybe (f x) i))
+                   → predTrans prog P i
+maybePTBindLemma {A} {m = m} {f} {P} {unit} prog refl nothingCase justCase
+   with runMaybe m unit | inspect (runMaybe m) unit
+... | nothing | [ R ] = predTrans-is-weakest m _
+      λ where r refl → subst (MaybebindPost _ _) (sym R) (nothingCase refl)
+... | just x  | [ R ]
+   with runMaybe (f x) unit | inspect (runMaybe (f x)) unit
+... | nothing | [ R2 ] = obm-dangerous-magic! "TODO: contradiction between R and R2"
```

Comment:
approved modulo finishing these TODOs.

Reply:
Thanks.  I'll keep working on eliminating the dangerous magic before merging and will request re-review when done.